libs/libc/stdio/lib_fputs.c

```diff
@@ -60,7 +60,7 @@
  ****************************************************************************/
 
 #if defined(CONFIG_ARCH_ROMGETC)
-int fputs(FAR const char *s, FAR FILE *stream)
+int fputs(FAR const IPTR char *s, FAR FILE *stream)
```

Comment:
I think we are miscommunicating.  The version of fouts should not be needed if IPTR is added.  It is needed only if IPTR is not present.

Reply:
Ok, I will remove the CONFIG_ARCH_ROMGETC, but if the compiler can generate the special instruction to read string from code region by IPTR, why we call up_romgetc for printf series?

Reply:
I don't know very much about how  IPTR works.  I had a different solution using up_romgetc and the IPTR changes were contributed later.  I have never used it 

We should both learn more before we change any code.

Reply:
Looking in include/nuttx/compiler.h, I see that IPTR is not necessarily supported for all AVRs:

    164 #elif defined(__AVR__)
    165 # if defined(CONFIG_AVR_HAS_MEMX_PTR)
    166   /* I-space access qualifiers needed by Harvard architecture */
    167
    168 #  define IOBJ __flash
    169 #  define IPTR __memx
    170
    171 # else
    172 /* No I-space access qualifiers */
    173
    174 #  define IOBJ
    175 #  define IPTR
    176 # endif

So, if __memx were not available, then yes you are right.  AVR would have to fall back and use the up_romgetc() functions.  Sorry about that.

Also the __memx qualifier is only available in the enhanced Atmel Studeo GCC compiler:

     63 config AVR_HAS_MEMX_PTR
     64         bool "Enable in-flash static const stings"
     65         depends on AVR_ATMEL_AVR_TOOLCHAIN
     66         select ARCH_DEBUG_H
     67         default y
     68         ---help---
     69                 Enabling this option (recommended) will place all constant
     70                 strings used for debugging and assertion in program memory
     71                 and allow the corresponding routines to get the strings
     72                 directly from there. This will dramatically decrease amount
     73                 of RAM needed to hold this static data.

If someone is using a generic AVR GCC compiler such as the one built by the NuttX buildroot, then __memx would not be available.

So you are right and I am wrong.  The alternative versions with up_romgetc() are needed in those cases.

     18 config ARCH_ROMGETC
     19         bool "Support for ROM string access"
     20         default n
     21         ---help---
     22                 In Harvard architectures, data accesses and instruction accesses
     23                 occur on different buses, perhaps concurrently.  All data accesses
     24                 are performed on the data bus unless special machine instructions
     25                 are used to read data from the instruction address space.  Also, in
     26                 the typical MCU, the available SRAM data memory is much smaller that
     27                 the non-volatile FLASH instruction memory.  So if the application
     28                 requires many constant strings, the only practical solution may be
     29                 to store those constant strings in FLASH memory where they can only
     30                 be accessed using architecture-specific machine instructions.
     31
     32                 If ARCH_ROMGETC is defined, then the architecture logic must export
     33                 the function up_romgetc().  up_romgetc() will simply read one byte
     34                 of data from the instruction space.
     35
     36                 If ARCH_ROMGETC is selected, certain C stdio functions are effected: (1)
     37                 All format strings in printf, fprintf, sprintf, etc. are assumed to lie
     38                 in FLASH (string arguments for %s are still assumed to reside in SRAM).
     39                 And (2), the string argument to puts and fputs is assumed to reside
     40                 in FLASH.  Clearly, these assumptions may have to modified for the
     41                 particular needs of your environment.  There is no "one-size-fits-all"
     42                 solution for this problem.

Reply:
One thing that always bothered me about this whole concept is that it assumes that all strings lie in FLASH.  Often, however, strings may be create in RAM via strcat or sprintf.  In that case fputs() similar functions would still fail.  There really is no good solution for AVR.

Reply:
The safest thing to do would be to make no changes now.  Or, perhaps as in your previous change, just add IPTR to the function prototype.  I think this is one of those subjects that requires and clear understanding of the effected platforms and a plan for exactly how we want them to work.  Let's not merge this until we have those things.

Reply:
Ok, I will restore to the previous one.

Reply:
> Ok, I will restore to the previous one.

I apologize.  I did not understand all of the issues.

Reply:
@patacongo let's finish this request.

Reply:
@xiaoxiang781216 Is there anything besides AVR that we support that uses this though?  For AVR I am not clear why `__memx` would not be defined.

Reply:
https://gcc.gnu.org/onlinedocs/gcc/Named-Address-Spaces.html shows `__mem` defined which is why I do not understand why it would not be in a "generic" avr gcc.

Reply:
I don't know that whether __mem become the standard gcc feature. But if it's true we can create a new patch to remove all up_romgetc stuff later in one patch. At least, we need make the current code ase consistent with each other: it's very strange that printf call up_romgetc but scanf don't.

Reply:
Ok. `__memx` looks like it has been part of gcc since 2012.